mpi-proxy-split/mpi-wrappers/mpi_collective_wrappers.cpp

```diff
@@ -170,27 +170,115 @@ USER_DEFINED_WRAPPER(int, Ibarrier, (MPI_Comm) comm, (MPI_Request *) request)
+   On the Reproducibility of MPI Reduction Operations
+
+ https://www.sciencedirect.com/science/article/pii/S0167819121000612
+   An optimisation of allreduce communication in message-passing systems
+
+ MPI standard:
+ Advice to users. Some applications may not be able to ignore the
+ non-associative nature of floating-point operations or may use
+ user-defined operations (see Section 5.9.5) that require a special
+ reduction order and cannot be treated as associative. Such applications
+ should enforce the order of evaluation explicitly. For example, in the
+ case of operations that require a strict left-to-right (or right-to-left)
+ evaluation order, this could be done by gathering all operands at a single
+ process (e.g., with MPI_GATHER), applying the reduction operation in the
+ desired order (e.g., with MPI_REDUCE_LOCAL), and if needed, broadcast or
+ scatter the result to the other processes (e.g., with MPI_BCAST). (End
+ of advice to users.)
+
+ And note that MPI_Waitany can receive messages non-determistically.
+ */
+#endif
+
+int
+MPI_Allreduce_reproducible(const void *sendbuf,
+                           void *recvbuf,
+                           int count,
+                           MPI_Datatype datatype,
+                           MPI_Op op,
+                           MPI_Comm comm,
+                           int dump_trace_from)
+{
+#define MAX_ALL_SENDBUF_SIZE (1024 * 1024 * 16) /* 15 MB */
+  // We use 'static' becuase we don't want the overhead of the compiler
+  // initializing these to zero each time the function is called.
+
+  static unsigned char tmpbuf[MAX_ALL_SENDBUF_SIZE];
+
+  int root = 0;
+  int comm_rank;
+  int comm_size;
+  int type_size;
+
+  MPI_Comm_rank(comm, &comm_rank);
+  MPI_Comm_size(comm, &comm_size);
+  MPI_Type_size(datatype, &type_size);
+
+  JASSERT(count * comm_size * type_size <= MAX_ALL_SENDBUF_SIZE);
+  JASSERT(sendbuf != FORTRAN_MPI_IN_PLACE);
+  JASSERT(sendbuf != MPI_IN_PLACE);
+
+  // Gather the operands from all ranks in the comm
+  MPI_Gather(sendbuf, count, datatype, tmpbuf, count, datatype, 0, comm);
+
+  // Perform the local reduction operation on the root rank
+  if (comm_rank == root) {
+    memset(recvbuf, 0, count * type_size);
+    memcpy(recvbuf, tmpbuf + (count * type_size * 0), count * type_size);
+    for (int i = 1; i < comm_size; i++) {
+      MPI_Reduce_local(tmpbuf + (count * type_size * i), recvbuf, count,
+                       datatype, op);
+    }
+  }
+
+  // Broadcat the local reduction operation result in the comm
+  int rc = MPI_Bcast(recvbuf, count, datatype, 0, comm);
+
+  return rc;
+}
+
 USER_DEFINED_WRAPPER(int, Allreduce,
                      (const void *) sendbuf, (void *) recvbuf,
                      (int) count, (MPI_Datatype) datatype,
                      (MPI_Op) op, (MPI_Comm) comm)
 {
+  char *s = getenv("PERFORM_REPRODUCIBLE_REDUCTION_OPERATION");
+  int perform_reproducible_reduction_operation = (s != NULL) ? atoi(s) : 0;
```

Comment:
In line with short, meaningful names, let's follow the naming of the environment variable, and use (here and everywhere):
`use_allreduce_reproducible`